---
title: "RAW_COUNTS_TO_TPM"
author: "Francesco Massaini"
date: "2024-03-18"
output: pdf_document
---

# Import and convert RAW counts into TPM. Starting w/ genes named with ENSEMBL ID 

# General info about how to import data
## Read tabular data into R
# read.table(file, header = FALSE, sep = "", dec = ".")

## Read "comma separated value" files (".csv")
# read.csv(file, header = TRUE, sep = ",", dec = ".", ...)

## Or use read.csv2: variant used in countries that 
## use a comma as decimal point and a semicolon as field separator.
# read.csv2(file, header = TRUE, sep = ";", dec = ",", ...)

## Read TAB delimited files
# read.delim(file, header = TRUE, sep = "\t", dec = ".", ...)
# read.delim2(file, header = TRUE, sep = "\t", dec = ",", ...)


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Imports

```{r Imports}
library(tidyverse)
library(tibble)
library(org.Hs.eg.db) # Homo sapiens db
library(ADImpute) # to normalize with gene lenght
library(RColorBrewer) # for plot
library(pheatmap) # for heatmaps
library(ADImpute) # This package provides functions for imputing missing data

```

# Functions 

```{r Functions}
# library(org.Hs.eg.db) # Homo sapiens db
# columns(org.Hs.eg.db) # to look at the columns of this db

# Raw counts to TPM
counts2TPM <- function(raw_counts_file){
  entrz <- AnnotationDbi::select(org.Hs.eg.db, keys = rownames(raw_counts_file), columns = "SYMBOL", keytype = "ENSEMBL") # keys are the data to overlap. columns is the column to replace and keytype is the column where to find corrispondences  
# entrz is a df storing ENSEMBL ID and GENE SYMBOLS. ENSEMBL ID are selected by the ROW NAMES of your raw_counts_file 
  TPM <- raw_counts_file %>%
    mutate(ENSEMBL = rownames(raw_counts_file)) %>% # add a column
    inner_join(., entrz, by="ENSEMBL") %>% # join, merge df with a common column (ENSEMBL)
    dplyr::filter(!is.na(SYMBOL)) %>%
    distinct(SYMBOL, .keep_all = T) %>% # keep only one variable of the duplicated symbols
    column_to_rownames("SYMBOL") %>% 
    dplyr::select(., -c("ENSEMBL"))

return(TPM)  
}

# Plots - Heatmap
TPM_Heatmap <- function(TPM){
  TPM <- as.matrix(TPM)
  x = ADImpute::NormalizeTPM(TPM, log = T)
  # Utilizes the NormalizeTPM function from the ADImpute package.
  # Performs normalization on the 'TPM' data frame, considering TPM (Transcripts Per Million) values.
  # Setting 'log = T' indicates that the normalization will be done using the logarithm of TPM values.
  # This is often employed to reduce variance and approximate a normal distribution.
  # Add 1 at the log log(TPM+1) so that 0 values are propertly calculated
  
  sampleDists <- dist(t(TPM)) #Compute distance of the matrix. DO NOT DO IT FOR GENES. dist take into account rows! So to look at the patients you need to transpose. Distance is computed with eucledian metric. Remember you need to transpose (t) the matrix because dist takes rows (and samples are into columns) 
  sampleDistMatrix <- as.matrix(sampleDists)
  colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
  pheatmap(sampleDistMatrix,
           clustering_distance_rows=sampleDists,
           clustering_distance_cols=sampleDists,
           col=colors,
           main="Heatmap between SAMPLES",
           fontsize= 20, fontsize_row = 5, fontsize_col = 5)
}

# Plots - Dendrogram
TPM_Sample_dendrogram <- function(TPM){
  sampleTree = hclust(dist(t(TPM)), method = "average");
  plot(sampleTree, 
       main = "Sample clustering to detect outliers", sub="", xlab="",
       cex.lab = 1.5, cex.axis = 1.5, cex.main = 2)
}
```




```{r Data Loading and plots}
# Data Loading

path <- c("/home/francesco.massaini/Desktop/IMMUCAN_Internship/BC2/BC2_COUNTS.txt",
          "/home/francesco.massaini/Desktop/IMMUCAN_Internship/BC3/BC3_COUNTS.txt",
          "/home/francesco.massaini/Desktop/IMMUCAN_Internship/NSCLC/NSCLC_COUNTS.txt")
files_names_ <- gsub(".txt$", "", basename(path))


# EDA
for (i in 1:length(path)) {
  counts <- read.table(path[i], header = TRUE,  sep = '	', row.names = 1)
  TPM <- counts2TPM(counts)
  write.table(TPM, file = paste0(files_names_[i],"_TPM"))
  pdf(paste0(files_names_[i],"_outliers"))
  TPM_Heatmap(TPM)
  TPM_Sample_dendrogram(TPM)
  dev.off()
}

```