---
output: html_document
editor_options: 
  chunk_output_type: inline
---
```{r}
library(ggplot2)
library(stringr)
library(dplyr)
getwd()
setwd(here::here())

source("../../../Assets/RNAseq_processing.R")

```

# merge RNA seq files and convert to TPM
```{r}
star_path <- '../../../IMMUcan_data/data/genomics/IMMUcan/STAR_count/'
# star_path <- './Projects/IMMUcan_data/data/genomics/IMMUcan/STAR_count/'
# filepattern <- "^IMMU.*\\.txt$"
filepattern <- "^LUNG\\-.*\\.txt$"

csv_files <- list.files(path = star_path, pattern = filepattern, full.names = TRUE) # 134


gene_expr <- combine_gene_expression_files(directory_path = star_path, separator = '\t', file_pattern = filepattern, sample_name_pattern = "LUNG-(.*)-[A-Z]+-[0-9]+-[A-Z]+-[0-9]+")

library(AnnotationDbi)
library(org.Hs.eg.db)

gene_ENS <- rownames(gene_expr)
gene_ENS <- gsub('(\\.[0-9]*)$','', gene_ENS)
rownames(gene_expr) <- gene_ENS

entrz <- AnnotationDbi::select(org.Hs.eg.db, keys = gene_ENS, columns = "SYMBOL", keytype = "ENSEMBL")

GeneCount <- gene_expr %>%
  data.frame() %>%
  mutate( ENSEMBL = rownames(gene_expr) ) %>%
  inner_join(., entrz, by="ENSEMBL")

GeneCount.selected <- select_from_duplicated_genes (GeneCount, gene_id_column = 'SYMBOL', selection_method = "mean")

symbols <- GeneCount.selected$SYMBOL

GeneCount.selected <- GeneCount.selected[,1:(length(GeneCount.selected)-2)]
rownames(GeneCount.selected) <- symbols


# TPM Normalization
Gene.count <- GeneCount.selected[,-ncol(GeneCount.selected)]
rownames(Gene.count) <- rownames(GeneCount.selected)
TPM.selected = ADImpute::NormalizeTPM(Gene.count, log=T)
TPM.selected <- as.data.frame(TPM.selected)

# write.table(as.data.frame(TPM.selected), '~/Projects/GEMDeCan_deconvolution/inputs/LUNG_tpm_120424.txt', sep = '\t')
```


```{r}
TPM.selected <- read.table('~/Projects/GEMDeCan_deconvolution/inputs/LUNG_tpm_120424.txt', sep = '\t', row.names = 1)
head(as.data.frame(TPM.selected))
```


# run EasieR

```{r}
library(easier)

hallmarks_of_immune_response <- c("CYT", "Ock_IS","IPS","IMPRES","MSI",
                                  "Roh_IS", "chemokines", "Davoli_IS", "IFNy", "Ayers_expIS", "Tcell_inflamed", "RIR", "TLS")


RNA_tpm<- read.delim("/home/francesco.massaini/Desktop/IMMUCAN_data/NSCLC2/02_TPM/NSCLC2_TPM.txt", row.names = 1)
immune_response_scores <- easier::compute_scores_immune_response	(RNA_tpm = RNA_tpm, 
                                                         selected_scores = hallmarks_of_immune_response)

head(immune_response_scores)

patients <- rownames(immune_response_scores)
x1 <- patients
x2 <- substr(x1, 6, 20)
x3 <- gsub('\\..*','',x2)
ann_col <- data.frame(Cohort = x3)
rownames(ann_col) <- x1
df_scaled <- as.data.frame(lapply(immune_response_scores, function(x) (x - min(x)) / (max(x) - min(x))))
rownames(df_scaled) <- rownames(immune_response_scores)

write.table(as.data.frame(immune_response_scores), '../../data/Intermediate_data/immune_response_scores_NSCLC2.tsv', sep = '\t')

# immune_response_scores <- read.table('~/Projects/GEMDeCan_deconvolution/inputs/immune_response_scores_NSCLC2.tsv', sep = '\t', row.names = 1)


pheatmap::pheatmap(df_scaled, annotation_row = ann_col, show_rownames = F)

```


```{r}
library(pheatmap)
df <- t(df_scaled[,c(-10,-11,-12,-13)])
pheatmap_result <- pheatmap(df)


# Perform hierarchical clustering manually
dist_matrix <- dist(t(df))  # Compute the distance matrix
hc <- hclust(dist_matrix)  # Perform hierarchical clustering
clusters <- cutree(hc, k = 5 )

D <- as.data.frame(clusters)
D$clusters <- as.character(D$clusters)

# To Edit Manually
D$clusters <- gsub( '1','Hot', D$clusters)
D$clusters <- gsub( '4','Intermediate Hot', D$clusters)
D$clusters <- gsub( '2','Intermediate Cold', D$clusters)
D$clusters <- gsub( '3','Cold', D$clusters)
D$clusters <- gsub( '5','Intermediate Cold', D$clusters)

ImmuneScores <- df_scaled[,c(-10,-11,-12,-13)]
pheatmap::pheatmap(t(df_scaled[,c(-10,-11,-12,-13)]), annotation_col = D, show_colnames = F)

ImmuneScores$ImmuneScores_Classification <- D$cluster

```

<!-- Distribution -->
```{r}
ScoreAvg <- data_frame(Score = rowMeans(df_scaled[,c(-10,-11,-12,-13)]))
rownames(ScoreAvg) <- row.names(df_scaled)
mrgd <- cbind(D, ScoreAvg)

mrgd[, "clusters"]<- ifelse(mrgd$Score <= .30, "Cold",
              ifelse(mrgd$Score >= .75, "Hot", "Intermediate"))


p <- ggplot2::ggplot(mrgd, aes(x = Score, color = clusters))+
  geom_density()+ theme_bw() + 
  geom_density(data = mrgd , color = 'black')+ theme_bw() + 
  
  labs(title=paste('Distribution of ImmuneScores NSCLC2' ))

print(p)
```


<!-- Load deconv -->

```{r}
IMMUcan_dec <- read.csv( '~/Projects/GEMDeCan_deconvolution/results/all_deconvolutions/all_deconvolutions_LUNG_NSCLC2_tpm_120424.txt', sep = '\t', row.names = 1)
source('~/Projects/Assets/RNAseq_processing.R')
IMMUcan_dec <- fix.names.GEMDeCan(IMMUcan_dec)

idx <- grepl("MCP",colnames(IMMUcan_dec))
# select NSCLC2
table(rownames(df_scaled) %in% rownames(IMMUcan_dec))

com <- rownames(df_scaled) %in% rownames(IMMUcan_dec)

IMMUcan_dec_NSCLC2 <- IMMUcan_dec[rownames(df_scaled)[com], !idx]

pheatmap::pheatmap(t(IMMUcan_dec_NSCLC2), annotation_col = D, show_colnames = F)

```

```{r}

source('~/Projects/Assets/RNAseq_processing.R')
# IMMUcan_dec <- fix.names.GEMDeCan(IMMUcan_dec)


IMMUcan_dec <- read.csv( '~/Projects/IMMUcan_Deliverables_Analysis/data/Intermediate_data/CTF/Deconvolution_after_subgrouping_NSCLC2.csv', row.names = 1)

idx <- grepl("MCP",colnames(IMMUcan_dec))
# select NSCLC2
rownames(IMMUcan_dec) <- gsub('\\-','\\.',rownames(IMMUcan_dec))

table(rownames(df_scaled) %in% rownames(IMMUcan_dec))

com <- rownames(df_scaled) %in% rownames(IMMUcan_dec)

IMMUcan_dec_NSCLC2 <- IMMUcan_dec[rownames(df_scaled)[com], !idx]

pheatmap::pheatmap(t(IMMUcan_dec_NSCLC2), annotation_col = D, show_colnames = F)

```



<!-- Significant deconvolution features -->
```{r}
common_samples <- rownames(df_scaled)[com]
features <- IMMUcan_dec_NSCLC2[common_samples,]
features$ImmuneScores_Classification <- D[common_samples,]

data = features[features$ImmuneScores_Classification %in% c('Hot','Cold'), ]
x_var <- 'ImmuneScores_Classification'
na_threshold = .5
Y_levels_closure = 10 # maximum number of different categorical values
Numeric_level_closure = 3 # min number of different numerical values
similarity_y_threshold = .95 # maximum allowed similiaty between values 
unknown_to_NA_ratio_threshold = .5 # maximum allowed Categorical variable when analysis Numerical features (replaced with NAs)
#output
source('~/Projects/Assets/stat_test.R')

results <- stat_test(data = data, x_var, na_threshold , Y_levels_closure, Numeric_level_closure = Numeric_level_closure, similarity_y_threshold = similarity_y_threshold)


pval_threshold <- 0.0005
significant_results <- results[results$p_value <= pval_threshold, ]

significant_vars <- unique(significant_results$y_var)

# Iterate over significant variables to create heatmaps
for (var in significant_vars) {
    if (! is.na(var)) {
      test_type <- results[results$y_var==var , ]$test
      if (test_type == 'ANOVA Numeric' ) {
        sign_features <- data[,c(var,x_var)]   
        colnames(sign_features) <- c('Feature','Category')
        # Generate the heatmap
        p <- ggplot2::ggplot(data = sign_features, aes(x = Category, y = Feature, fill = Category)) +
          geom_violin(alpha=.3)+
          geom_point()+
          labs(x = x_var, y = var)+
          theme_bw()
        print(p)
    }
  }
}

```

```{r}
data$ImmuneScores_Classification
data$samples <-rownames(data) 
melted_df <- reshape2::melt( id.vars = c('samples','ImmuneScores_Classification') , data)
reconstructed_df <- reshape2::dcast(melted_df, ImmuneScores_Classification~variable, mean , value.var = 'value')
rownames(reconstructed_df) <- reconstructed_df$ImmuneScores_Classification
reconstructed_df <- reconstructed_df[,-1]


method_sig_ct <- gsub('Quantiseq_','Quantiseq_Defeault_', colnames(reconstructed_df))
method_sig_ct <- as.data.frame(stringr::str_split_fixed(method_sig_ct, "_", 3))
colnames(method_sig_ct) <- c('Method','Signature','Specific.celltype')
method_sig_ct$rownames <- colnames(reconstructed_df)

dict <- read.csv2('/home/abdelmounim.essabbar/Projects/ITI_NSCLC/data/intemediate/dictionary_celltypes.csv', sep='\t')
method_sig_ct <- merge(method_sig_ct, dict, all.x = TRUE)
rownames(method_sig_ct) <- method_sig_ct$rownames
method_sig_ct$Specific.celltype
pheatmap(reconstructed_df, annotation_col = method_sig_ct[,c('Cell.type'),drop=F], show_colnames = T )

method_sig_ct <- method_sig_ct[,-4]


```

```{r}

for ( CT in unique(method_sig_ct$Cell.type)) {
  idx <- method_sig_ct$Cell.type == CT
  method_ct_selected <-  na.omit(method_sig_ct[idx, ])
  selected_methods <- rownames(method_ct_selected)
  if (length(selected_methods)>1) {
    matrix_by_ct <- reconstructed_df[,selected_methods]
  }
  pheatmap(matrix_by_ct, annotation_col = method_sig_ct[,,drop=F] )
}
```


###########

```{r}

CT <- 'T'
idx <- method_sig_ct$Cell.type == CT
method_ct_selected <-  na.omit(method_sig_ct[idx, ])
selected_methods <- rownames(method_ct_selected)
if (length(selected_methods)>1) {
  matrix_by_ct <- reconstructed_df[,selected_methods]
  pheatmap(matrix_by_ct, annotation_col = method_sig_ct[,,drop=F] )
}

```



<!-- Correlation between immunescores and DeconvolutionFeatures -->
```{r}
IMMUcan_dec_NSCLC <- IMMUcan_dec_NSCLC2
dim(ImmuneScores)
IMMUcan_dec_NSCLC$immucan_id <- gsub('\\.','-', row.names(IMMUcan_dec_NSCLC) )
IMMUcan_dec_NSCLC$immucan_id <- substr(IMMUcan_dec_NSCLC$immucan_id, 1,16  )
IS_deconv <- merge(IMMUcan_dec_NSCLC, ImmuneScores, on = 'immucan_id')
IS_deconv <- IS_deconv[,c(-1, - ncol(IS_deconv))]

Co <- cor(IS_deconv)
CO <- Co[87:96,1:86]

co <- na.omit(CO)

method_sig_ct <- colnames(co)
# table(grepl('Quantiseq',method_sig_ct))
method_sig_ct <- gsub('Quantiseq_','Quantiseq_Defeault_', method_sig_ct)
method_sig_ct <- as.data.frame(stringr::str_split_fixed(method_sig_ct, "_", 3))
colnames(method_sig_ct) <- c('Method','Signature','Specific.celltype')
method_sig_ct$rownames <- colnames(IMMUcan_dec_NSCLC)

dict <- read.csv2('/home/abdelmounim.essabbar/Projects/ITI_NSCLC/data/intemediate/dictionary_celltypes.csv', sep='\t')
method_sig_ct <- merge(method_sig_ct, dict, all.x = TRUE)
rownames(method_sig_ct) <- method_sig_ct$rownames

library(RColorBrewer)
dim(co)
coul <- colorRampPalette(brewer.pal(10, "RdBu") )(25)
pheatmap(co, col = rev(coul),  main="Correlation plot", show_rownames = F, annotation_row = method_sig_ct[rownames(co),c(-1,-3,-4)])

```


```{r}
IMMUcan_dec <- read.csv( '~/Projects/GEMDeCan_deconvolution/results/all_deconvolutions/all_deconvolutions_LUNG_NSCLC2_tpm_120424.txt', sep = '\t', row.names = 1)
source('~/Projects/Assets/RNAseq_processing.R')


IMMUcan_dec <- read.csv( '~/Projects/IMMUcan_Deliverables_Analysis/data/Intermediate_data/CTF/Deconvolution_after_subgrouping_NSCLC2.csv', row.names = 1)

idx <- grepl("MCP",colnames(IMMUcan_dec))
# select NSCLC2
rownames(IMMUcan_dec) <- gsub('\\-','\\.',rownames(IMMUcan_dec))

table(rownames(df_scaled) %in% rownames(IMMUcan_dec))

com <- rownames(df_scaled) %in% rownames(IMMUcan_dec)

IMMUcan_dec_NSCLC2 <- IMMUcan_dec[rownames(df_scaled)[com], !idx]

pheatmap::pheatmap(t(IMMUcan_dec_NSCLC2), annotation_col = D, show_colnames = F)
```


```{r}

common_samples <- rownames(df_scaled)[com]
features <- IMMUcan_dec_NSCLC2[common_samples,]
features$ImmuneScores_Classification <- D[common_samples,]

features <- features[features$ImmuneScores_Classification %in% c('Hot','Cold'),]

data = features
x_var <- 'ImmuneScores_Classification'
na_threshold = .5
Y_levels_closure = 10 # maximum number of different categorical values
Numeric_level_closure = 3 # min number of different numerical values
similarity_y_threshold = .95 # maximum allowed similiaty between values 
unknown_to_NA_ratio_threshold = .5 # maximum allowed Categorical variable when analysis Numerical features (replaced with NAs)
#output
source('~/Projects/Assets/stat_test.R')

results <- stat_test(data = data, x_var, na_threshold , Y_levels_closure, Numeric_level_closure = Numeric_level_closure, similarity_y_threshold = similarity_y_threshold)


pval_threshold <- 0.05
significant_results <- results[results$p_value <= pval_threshold, ]

significant_vars <- unique(significant_results$y_var)
```


```{r}
# Iterate over significant variables to create heatmaps
for (var in significant_vars) {
    if (! is.na(var)) {
      test_type <- results[results$y_var==var , ]$test
      if (test_type == 'ANOVA Numeric' | test_type == 'T-Test' ) {
        sign_features <- data[,c(var,x_var)]   
        colnames(sign_features) <- c('Feature','Category')
        # Generate the heatmap
        p <- ggplot2::ggplot(data = sign_features, aes(x = Category, y = Feature, fill = Category)) +
          geom_violin(alpha=.8)+
          geom_point()+
          geom_jitter(size=.2)+
          labs(x = x_var, y = var)+
          scale_fill_manual(values=c("navy", "orange", "#56B4E9")) + 
          theme_bw()
        p + labs(title = paste('Test:',results$test, 'pval:', format(results$p_value, scientific = TRUE),'\n', results$x_var, '&' , results$y_var ,' ' ))
        print(p)
    }
  }
}

```






<!-- Correlation with Imaging features -->
```{r}
#IF1
IF1 <- read.csv('/home/abdelmounim.essabbar/Projects/IMMUcan_data/from_daniel/IF1_cell_type_fraction.csv',row.names = 1)
colnames(IF1) <- paste0('IF1_',colnames(IF1))
IF1$sample_id <- substr(rownames(IF1),1,24)
#IF2
IF2 <- read.csv('/home/abdelmounim.essabbar/Projects/IMMUcan_data/from_daniel/IF2_cell_type_fraction.csv',row.names = 1)
colnames(IF2) <- paste0('IF2_',colnames(IF2))
IF2$sample_id <- substr(rownames(IF2),1,24)
#IF3
IF3 <- read.csv('/home/abdelmounim.essabbar/Projects/IMMUcan_data/from_daniel/IF3_cell_type_fraction.csv',row.names = 1)
colnames(IF3) <- paste0('IF3_',colnames(IF3))
IF3$sample_id <- substr(rownames(IF3),1,24)

#Merging IFs
IFs <- merge(IF1, IF2, by="sample_id")
IFs <- merge(IFs, IF3, by="sample_id")
rownames(IFs) <- IFs$sample_id
IFs$immucan_id <- substr(rownames(IFs),1,16)

imaging_features <- colnames(IFs)[c(-1,-ncol(IFs)) ]

ImmuneScores$immucan_id <- gsub('\\.','\\-', substr(rownames(ImmuneScores), 1, 16) )

write.csv(ImmuneScores, './../../data/Intermediate_data/ImmuneScores_NSCLC2_groups.csv' )
```


```{r}
Immune.Scores <- ImmuneScores[,c('ImmuneScores_Classification','immucan_id'),drop=F]
Img_features <- merge(Immune.Scores, IFs[,-1])[,-1]


# Perform statistic test

data = Img_features[Img_features$ImmuneScores_Classification %in% c('Hot','Cold'), ]
groups <- unique(data[, x_var])

x_var <- 'ImmuneScores_Classification'
na_threshold = .5
Y_levels_closure = 10 # maximum number of different categorical values
Numeric_level_closure = 3 # min number of different numerical values
similarity_y_threshold = .95 # maximum allowed similiaty between values 
unknown_to_NA_ratio_threshold = .5 # maximum allowed Categorical variable when analysis Numerical features (replaced with NAs)
#output
source('~/Projects/Assets/stat_test.R')

results <- stat_test(data = data, x_var, na_threshold , Y_levels_closure, Numeric_level_closure = Numeric_level_closure, similarity_y_threshold = similarity_y_threshold)

```









<!-- Plots -->
```{r}

pval_threshold <- 0.05
significant_results <- results[results$p_value <= pval_threshold, ]

significant_vars <- unique(significant_results$y_var)

# Iterate over significant variables to create heatmaps
for (var in significant_vars) {
    if (! is.na(var)) {
      test_type <- results[results$y_var==var , ]$test
      if (test_type == 'ANOVA Numeric' | test_type == 'T-Test' ) {
        sign_features <- data[,c(var,x_var)]   
        colnames(sign_features) <- c('Feature','Category')
        # Generate the heatmap
        p <- ggplot2::ggplot(data = sign_features, aes(x = Category, y = Feature, fill = Category)) +
          geom_violin(alpha=.8)+
          geom_point()+
          geom_jitter(size=.2)+
          labs(x = x_var, y = var, title = paste(test_type , '\t', var , '&', x_var) )+
          scale_fill_manual(values=c("lightblue1", "salmon1", "#56B4E9")) + 
          theme_bw()
        p + labs(title = paste('Test:',results$test, 'pval:', format(results$p_value, scientific = TRUE),'\n', results$x_var, '&' , results$y_var ,' ' ))
        print(p)
    }
  }
}





```




```{r}
y_var = 'IF1_Tumor_fraction'
data$ImmuneScores_Classification
ggplot2::ggplot(data = data, aes(x = ImmuneScores_Classification, y = IF1_Tumor_fraction, fill = ImmuneScores_Classification)) +
          geom_violin(alpha=.8)+
          geom_point()+
          geom_jitter(size=.2)+
          # labs(x = x_var, y = y_var, title = paste(test_type , '\t', var , '&', x_var) )+
          scale_fill_manual(values=c("lightblue1", "salmon1", "#56B4E9")) + 
          theme_bw()
```


<!-- Hot/Cold and clinical data -->
```{r}
library(rstatix)
clin <- read.csv('/home/abdelmounim.essabbar/Projects/IMMUcan_data/from_daniel/NSCLC2_for_R_update_2024.csv')
clin$immucan_id
ImmuneScores$immucan_sample_id <- gsub('\\.','\\-', substr(rownames(ImmuneScores), 1, 24) )
ImmuneScores$immucan_id <- gsub('\\.','\\-', substr(rownames(ImmuneScores), 1, 16) )

ImmuneScores$ImmuneScores_Classification
com <- clin[ clin$immucan_id %in% ImmuneScores$immucan_id, ]$immucan_id
rownames(clin) <- clin$immucan_id
rownames(ImmuneScores) <- ImmuneScores$immucan_id


clin_s <- clin[com,]
ImmuneScores_s <- ImmuneScores[com,]
Clinical_features <- cbind(clin_s , ImmuneScores_s[,'ImmuneScores_Classification', drop=F])

FoI <- c('age_at_registration','age_at_diagnosis','gender','average_number_of_cigarettes','number_of_years','smoking_status','death','time_to_event','curative_surgery','adjuvant_treatment','type_of_adj_trt','metastatic_treatment','simple_histology','simple_stage','long_survivors','T_stage','tumor_size_integer','TIL_score','tumor_area_mm_square','ImmuneScores_Classification') # Features of Interest

Clinical_features <- Clinical_features[,FoI]

cols_with_variability <- sapply(Clinical_features, function(x) length(unique(x)) > 1)
Clinical_features <- Clinical_features[,cols_with_variability]

head(Clinical_features)
```


```{r}
data = Clinical_features
x_var <- 'ImmuneScores_Classification'
na_threshold = .5
Y_levels_closure = 10 # maximum number of different categorical values
Numeric_level_closure = 3 # min number of different numerical values
similarity_y_threshold = .95 # maximum allowed similiaty between values 
unknown_to_NA_ratio_threshold = .5 # maximum allowed Categorical variable when analysis Numerical features (replaced with NAs)
#output
results <- data.frame(variable = character(), p_value = numeric(), stat = numeric(), stringsAsFactors = FALSE)
source('../../../Assets/stat_test.R')

data <- data[data$ImmuneScores_Classification %in% c('Hot','Cold'), ]
results <- stat_test(data = data, x_var, na_threshold , Y_levels_closure, Numeric_level_closure = Numeric_level_closure, similarity_y_threshold = similarity_y_threshold)
```

```{r}
pval_threshold <- 0.1
significant_results <- results[results$p_value <= pval_threshold, ]

significant_vars <- unique(significant_results$y_var)


# Iterate over significant variables to create heatmaps
for (var in significant_vars) {
    if (! is.na(var)) {
      test_type <- results[results$y_var==var , ]$test
      p_value <- results[results$y_var==var , ]$p_value
      X <- results[results$y_var==var , ]$x_var
      Y <- results[results$y_var==var , ]$y_var
      if (test_type == 'ANOVA Numeric' | test_type == 'T-Test' ) {
        sign_features <- data[,c(var,x_var)]   
        colnames(sign_features) <- c('Feature','Category')
        # Generate the heatmap
        p <- ggplot2::ggplot(data = sign_features, aes(x = Category, y = Feature, fill = Category)) +
          geom_violin(alpha=.8)+
          geom_point()+
          geom_jitter(size=.2)+
          labs(x = x_var, y = var, title = paste(test_type , '\t', var , '&', x_var) )+
          scale_fill_manual(values=c("darkolivegreen2", "brown3", "#56B4E9")) + 
          theme_bw()+ labs(title = paste('Test:',test_type, 'pval:', format(p_value, scientific=T ), '\n', X, '&' , Y ,' ' ))
        print(p)
      }
      else if ( test_type == 'ChiSquare' ) {
          # Create a cross table
          crosstab <- table(data[,var], data[,x_var])
          
          # Convert the cross table into a long format data frame suitable for ggplot2
          crosstab_melted <- reshape2::melt(crosstab)
          
          # Generate the heatmap
          p <- ggplot(data = crosstab_melted, aes(x = Var1, y = Var2, fill = value)) +
            geom_tile(color = "black") +
            scale_fill_gradient(low = "lightblue1", high = "red2") +
            labs(title = paste("Heatmap of", var, "vs.",x_var),
                 x = var,
                 y = x_var,
                 fill = "Count") +
            theme_minimal() +
            theme(axis.text.x = element_text(angle = 45, hjust = 1))
          print(p)

      }
      
  }
}


```

